### PR TITLE
feat: Update wasmer dependency to allow for stable toolchain for nearcore

### PR DESCRIPTION
**Pull request by willemneal**
I published the updated wasmer dependencies to crates.io so that this can now be published too.

@evgenykuzyakov can I just publish near-vm-runner to crates.io?

### CODE REVIEW

**Comment by willemneal** [comment]
It seems that lightbeam uses `staticvec`, which requires nightly.  Since it's an optional dependency is it okay to make only tests that involve it use nightly and keep the top level toolchain file?

**Review by ailisp — APPROVED** [review]
> It seems that lightbeam uses staticvec, which requires nightly. Since it's an optional dependency is it okay to make only tests that involve it use nightly and keep the top level toolchain file?

I think so. took a look at staticvec, it use constant generics, and the progress doesn't look like something that will go into stable soon: https://github.com/rust-lang/rust/issues/44580. So I think it not worth to wait it to make whole nearcore on stable rust

**Comment by olonho** [comment]
Seems not all integration tests are green yet, could you please fix it?

**Comment by willemneal** [comment]
Putting this here for others as I mentioned this to you privately: the last hurdle to making the integration tests pass is the the param estimator's use of `llvm_asm`, which is a nightly feature.  `dynasm` should allow the same functionality but is stable.  Hopefully we can fix this at our meeting tomorrow.

File: chain/rosetta-rpc/Cargo.toml [hunk 1]
```diff
@@ -22,7 +22,7 @@ serde = { version = "1", features = ["derive"] }
 serde_json = "1"
 validator = "0.10"
 
-paperclip = { version = "0.4", features = ["actix", "actix-nightly"] }
+paperclip = { version = "0.4", features = ["actix"] }
```
Comment (frol) [inline chain/rosetta-rpc/Cargo.toml:25]: is it still compilable with `rosetta_rpc` feature enabled? (the feature is disabled by default)

```
cargo check --features rosetta_rpc
```

File: runtime/runtime-params-estimator/Cargo.toml [hunk 2]
```diff
@@ -28,8 +28,10 @@ neard = { path = "../../neard" }
 rocksdb = { git = "https://github.com/nearprotocol/rust-rocksdb", branch="disable-thread" }
 glob = "0.3.0"
 walrus = "0.18.0"
+dynasm = "1.0.0"
```
Comment (frol) [inline runtime/runtime-params-estimator/Cargo.toml:31]: It seems that you use a re-exported version from `dynasmrt`, do you still need it as an explicit dependency?

File: runtime/near-vm-runner/Cargo.toml [hunk 3]
```diff
@@ -14,11 +14,11 @@ This crate implements the specification of the interface that Near blockchain ex
...
 [dependencies]
 cached = "0.12"
-wasmer-runtime = { version = "=0.17.1", features = ["default-backend-singlepass"], default-features = false }
-wasmer-runtime-core = { version = "=0.17.1" }
+wasmer-runtime = { version="0.17.1", features = ["default-backend-singlepass"], default-features = false, package = "wasmer-runtime-near" }
+wasmer-runtime-core = {version = "0.17.1", package = "wasmer-runtime-core-near" }
 pwasm-utils = "0.12"
 parity-wasm = "0.41"
-wasmtime = { version = "0.17.0", features = ["lightbeam"], default-features = false, optional = true }
+wasmtime = { version = "0.20.0", default-features = true, optional = true }
```
Comment (frol) [inline runtime/near-vm-runner/Cargo.toml:21]: Why did you enable `default-features`? It seems that it pulls more than we need.

**Comment by frol** [comment]
Once CI is green, I suggest you run nayduck for this PR before merging.

File: runtime/runtime-params-estimator/Cargo.toml (hunk 2 above)
Comment (willemneal) [reply runtime/runtime-params-estimator/Cargo.toml]: Found a way around it, now just use rust's file read with a raw fd, which the generated assembly was doing with a syscall.

File: runtime/near-vm-runner/Cargo.toml (hunk 3 above)
Comment (willemneal) [reply runtime/near-vm-runner/Cargo.toml:21]: Oops!

File: chain/rosetta-rpc/Cargo.toml (hunk 1 above)
Comment (willemneal) [reply chain/rosetta-rpc/Cargo.toml:25]: Yep, do you know what was needed by `actix-nightly`?

Comment (frol) [reply chain/rosetta-rpc/Cargo.toml:25]: I believe I was getting some cryptic errors and I was trying to use nightly features to get better error messages from macros. If it is compilable without it, it is fine to drop it :+1:

File: runtime/runtime-params-estimator/src/testbed_runners.rs [hunk 4]
```diff
@@ -108,18 +109,16 @@ pub fn measure_actions(
 }
 
 // TODO: super-ugly, can achieve the same via higher-level wrappers over POSIX read().
```
Comment (frol) [inline runtime/runtime-params-estimator/src/testbed_runners.rs:111]: It seems that this TODO is getting addressed. @olonho Can you review this implementation? (according to git, it was your TODO)

File: runtime/runtime-params-estimator/Cargo.toml [hunk 5]
```diff
@@ -32,4 +32,4 @@ walrus = "0.18.0"
 [features]
 default = []
 wasmtime = ["near-vm-logic/wasmtime_default"]
-lightbeam = ["wasmtime", "near-vm-runner/lightbeam"]
+#lightbeam = ["wasmtime", "near-vm-runner/lightbeam"]
```
Comment (frol) [inline runtime/runtime-params-estimator/Cargo.toml:35]: friendly reminder: clean up unused dependencies

**Review by frol — APPROVED** [review]

File: runtime/runtime-params-estimator/src/testbed_runners.rs (hunk 4 above)
Comment (willemneal) [reply runtime/runtime-params-estimator/src/testbed_runners.rs:111]: He's OOO today, but I've spoken to him about this solution and he has ideas on how to improve it.

File: runtime/runtime-params-estimator/src/testbed_runners.rs [hunk 6]
```diff
@@ -108,18 +109,15 @@ pub fn measure_actions(
...
-#[cfg(any(target_arch = "x86_64"))]
+#[cfg(target_family = "unix")]
```
Comment (olonho) [inline runtime/runtime-params-estimator/src/testbed_runners.rs:112]: Please make this a PR on its own and let's commit it independently of migration to stable Rust, as it is valuable anyway.

Comment (willemneal) [reply runtime/runtime-params-estimator/src/testbed_runners.rs:112]: Do you mean just this line or the change to use file read?

**Comment by willemneal** [comment]
@frol I saw that you had edited `deny.toml` recently and was wondering if you had any ideas on how to fix the current errors I'm getting.

**Comment by frol** [comment]
@willemneal how did you update `Cargo.lock`? Try reverting Cargo.lock to the version that is currently in `master` and just run `cargo check --workspace`. That should update only the relevant parts of Cargo.lock. We have not started upgrading our dependencies as we have a huge list of dependencies, and it is hard to make sure we don't accidentally break something.

**Comment by frol** [comment]
Well, it seems there is no other way around but to whitelist the duplicate crates (good news is that we can remove some other whitelisted items since they got deduplicated through the upgrade)

**Comment by willemneal** [comment]
@frol, locally the ban check passes, but it still fails CI.  Any thoughts?

File: runtime/near-vm-runner/Cargo.toml [hunk 7]
```diff
@@ -14,11 +14,11 @@ This crate implements the specification of the interface that Near blockchain ex
 
 [dependencies]
 borsh = "0.7.1"
-wasmer-runtime = { version = "=0.17.1", features = ["default-backend-singlepass"], default-features = false }
-wasmer-runtime-core = { version = "=0.17.1" }
+wasmer-runtime = { version="0.17.1", features = ["default-backend-singlepass"], default-features = false, package = "wasmer-runtime-near" }
+wasmer-runtime-core = {version = "0.17.1", package = "wasmer-runtime-core-near" }
 pwasm-utils = "0.12"
 parity-wasm = "0.41"
-wasmtime = { version = "0.17.0", features = ["lightbeam"], default-features = false, optional = true }
```
Comment (olonho) [inline runtime/near-vm-runner/Cargo.toml:21]: Why this change? We do need lightbeam, IMO.

File: runtime/runtime-params-estimator/src/testbed_runners.rs [hunk 8]
```diff
@@ -109,18 +110,15 @@ pub fn measure_actions(
...
 #[inline(always)]
-pub unsafe fn syscall3(mut n: usize, a1: usize, a2: usize, a3: usize) -> usize {
-    llvm_asm!("syscall"
-         : "+{rax}"(n)
-         : "{rdi}"(a1) "{rsi}"(a2) "{rdx}"(a3)
-         : "rcx", "r11", "memory"
-         : "volatile");
-    n
+pub unsafe fn syscall3(fd: u32, buf: &mut [u8]) {
```
Comment (olonho) [inline runtime/runtime-params-estimator/src/testbed_runners.rs:115]: Think we can do that simpler, like is being done in https://github.com/near/nearcore/pull/3705/files#diff-476f808c2693b5c35256c89880c018915744db6a28359ef8132fcb6d5184b2ac

File: rust-toolchain [hunk 9]
```diff
@@ -1 +1 @@
-nightly-2020-05-15
+nightly-2020-06-03
```
Comment (olonho) [inline rust-toolchain:1]: We we move to different nightly?

File: runtime/near-vm-runner/Cargo.toml (hunk 7 above)
Comment (willemneal) [reply runtime/near-vm-runner/Cargo.toml:21]: Oh I thought we had discussed this and since there isn't really any clear development on it we should just remove it.  Also it's not stable so I removed it for this reason.  I can investigate it further if you think we should keep it around.

File: rust-toolchain (hunk 9 above)
Comment (willemneal) [reply rust-toolchain:1]: Yeah it was needed for dynasm.

File: runtime/runtime-params-estimator/src/testbed_runners.rs (hunk 8 above)
Comment (willemneal) [reply runtime/runtime-params-estimator/src/testbed_runners.rs:115]: That looks good.  Since you are updating the API in that PR, is it okay to just merge this as is since it works and then update  with your PR?

Comment (willemneal) [reply runtime/runtime-params-estimator/src/testbed_runners.rs]: NVM I updated this PR.

File: runtime/near-vm-runner/Cargo.toml (hunk 7 above)
Comment (willemneal) [reply runtime/near-vm-runner/Cargo.toml:21]: So I added it back, but we need to still use nightly to compile it.  If we could just have it be ignored when doing the top level check with `--all-features` then everything would be fine.  I guess we'll have to perhaps move the param estimator outside of the workspace so that that check doesn't try this feature when compiling with stable.  Otherwise I think it's fine since the downstream deps like the sdk can still build with stable.

**Review by olonho — APPROVED** [review]
